File: problems/CSES/Graph_Algorithms/high_score/high_score.py
import sys
from collections import deque
# sys.setrecursionlimit(200000)

input = sys.stdin.readline

def main():
    n, m = map(int, input().split())

    edges = []
    reverse_graph = [[] for _ in range(n)]

    for _ in range(m):
        a, b, x = map(int, input().split())
        a -= 1
        b -= 1

        edges.append((a, b, x))
        reverse_graph[b].append(a)

    can_reach_n = [False] * n
    queue = deque([n - 1])
    can_reach_n[n - 1] = True

    while queue:
        u = queue.popleft()

        for v in reverse_graph[u]:
            if not can_reach_n[v]:
                can_reach_n[v] = True
                queue.append(v)

    NEG_INF = -10 ** 30

    dist = [NEG_INF] * n
    dist[0] = 0

    # Bellman-Ford
    for _ in range(n - 1):
        changed = False

        for u, v, w in edges:
            if dist[u] == NEG_INF:
                continue

            if dist[u] + w > dist[v]:
                dist[v] = dist[u] + w
                changed = True

        if not changed:
            break

    for u, v, w in edges:
        if dist[u] == NEG_INF:
            continue

        if dist[u] + w > dist[v]:
            if can_reach_n[v]:
                print(-1)
                return

    print(dist[n - 1])

main()

File: problems/CSES/Graph_Algorithms/high_score/high_score.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
#define pb push_back

struct Edge {
    int u, v;
    ll w;
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, m;
    cin >> n >> m;

    vector<Edge> edges;
    vector<vector<int>> reverse_graph(n);

    for (int i = 0; i < m; i++) {
        int a, b;
        ll x;

        cin >> a >> b >> x;

        a--;
        b--;

        edges.push_back({a, b, x});
        reverse_graph[b].push_back(a);
    }

    vector<bool> can_reach_n(n, false);

    queue<int> q;
    q.push(n - 1);
    can_reach_n[n - 1] = true;

    while (!q.empty()) {
        int u = q.front();
        q.pop();

        for (int v : reverse_graph[u]) {
            if (!can_reach_n[v]) {
                can_reach_n[v] = true;
                q.push(v);
            }
        }
    }

    const ll NEG_INF = -(1LL << 60);

    vector<ll> dist(n, NEG_INF);
    dist[0] = 0;

    for (int i = 0; i < n - 1; i++) {
        bool changed = false;

        for (auto [u, v, w] : edges) {
            if (dist[u] == NEG_INF)
                continue;

            if (dist[u] + w > dist[v]) {
                dist[v] = dist[u] + w;
                changed = true;
            }
        }

        if (!changed)
            break;
    }

    for (auto [u, v, w] : edges) {
        if (dist[u] == NEG_INF)
            continue;

        if (dist[u] + w > dist[v]) {
            if (can_reach_n[v]) {
                cout << -1 << '\n';
                return 0;
            }
        }
    }

    cout << dist[n - 1] << '\n';

    return 0;
}
